be/src/exec/pipeline/fragment_executor.cpp

```diff
@@ -701,10 +701,12 @@ Status FragmentExecutor::execute(ExecEnv* exec_env) {
     return Status::OK();
 }
 
-void FragmentExecutor::_fail_cleanup() {
+void FragmentExecutor::_fail_cleanup(bool fragment_has_registed) {
     if (_query_ctx) {
         if (_fragment_ctx) {
-            _query_ctx->fragment_mgr()->unregister(_fragment_ctx->fragment_instance_id());
+            if (fragment_has_registed) {
+                _query_ctx->fragment_mgr()->unregister(_fragment_ctx->fragment_instance_id());
+            }
             _fragment_ctx->destroy_pass_through_chunk_buffer();
             _fragment_ctx.reset();
         }
```

Comment:
`count_down_fragments` should also be invoked conditionally?

It is incremented in `QueryContextManager::get_or_register` in `_prepare_query_ctx`.

```cpp
    if (_query_ctx) {
        if (_fragment_ctx) {           
             if (fragment_has_registed) {
                _query_ctx->fragment_mgr()->unregister(_fragment_ctx->fragment_instance_id());
             }
            _fragment_ctx->destroy_pass_through_chunk_buffer();
            _fragment_ctx.reset();
        }
        if (fragment_has_incremented) {
            _query_ctx->count_down_fragments();
        }
    }
```

Reply:
if query_ctx not null. it means fragment has incremented
```
    _query_ctx = exec_env->query_context_mgr()->get_or_register(query_id);
```
```
QueryContext* QueryContextManager::get_or_register(const TUniqueId& query_id) {
    size_t i = _slot_idx(query_id);
    auto& mutex = _mutexes[i];
    auto& context_map = _context_maps[i];
    auto& sc_map = _second_chance_maps[i];

    {
        std::shared_lock<std::shared_mutex> read_lock(mutex);
        // lookup query context in context_map
        auto it = context_map.find(query_id);
        if (it != context_map.end()) {
            it->second->increment_num_fragments();
            return it->second.get();
        }
    }
...
```

Reply:
I see.